activestorage/lib/active_storage/analyzer/video_analyzer.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/hash/compact"
+
+module ActiveStorage
+  # Extracts the following from a video blob:
+  #
+  # * Width (pixels)
+  # * Height (pixels)
+  # * Duration (seconds)
+  # * Angle (degrees)
+  # * Aspect ratio
+  #
+  # Example:
+  #
+  #   ActiveStorage::VideoAnalyzer.new(blob).metadata
+  #   # => { width: 640, height: 480, duration: 5.0, angle: 0, aspect_ratio: [4, 3] }
+  #
+  # This analyzer requires the {ffmpeg}[https://www.ffmpeg.org] system library, which is not provided by Rails. You must
+  # install ffmpeg yourself to use this analyzer.
+  class Analyzer::VideoAnalyzer < Analyzer
+    def self.accept?(blob)
+      blob.video?
+    end
+
+    def metadata
+      { width: width, height: height, duration: duration, angle: angle, aspect_ratio: aspect_ratio }.compact
+    end
+
+    private
+      def width
+        Integer(video_stream["width"]) if video_stream["width"]
+      end
+
+      def height
+        Integer(video_stream["height"]) if video_stream["height"]
+      end
+
+      def duration
+        Float(video_stream["duration"]) if video_stream["duration"]
+      end
+
+      def angle
+        Integer(tags["rotate"]) if tags["rotate"]
+      end
+
+      def aspect_ratio
+        if descriptor = video_stream["display_aspect_ratio"]
+          descriptor.split(":", 2).collect(&:to_i)
+        end
+      end
+
+
+      def tags
+        @tags ||= video_stream["tags"] || {}
+      end
+
+      def video_stream
+        @video_stream ||= streams.detect { |stream| stream["codec_type"] == "video" } || {}
+      end
+
+      def streams
+        probe["streams"] || []
+      end
+
+      def probe
+        download_blob_to_tempfile { |file| probe_from(file) }
+      end
+
+      def probe_from(file)
+        IO.popen([ "ffprobe", "-print_format", "json", "-show_streams", "-v", "error", file.path ]) do |output|
```

Comment:
what happens if the machines doesn't have `ffprobe`?

Reply:
An `Errno::ENOENT` exception would have been raised. FFmpeg commonly won’t be installed at first, so I switched to logging an informative message and failing gracefully in 1d4b6db.

Reply:
:+1: